base/compiler/typeinfer.jl

```diff
@@ -8,7 +8,192 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
     return typeinf(interp, frame)
 end
 
+"""
+The module `Core.Compiler.Timings` provides a simple implementation of nested timers that
+can be used to measure the exclusive time spent inferring each method instance that is
+recursively inferred during type inference.
+
+This is meant to be internal to the compiler, and makes some specific assumptions about
+being used for this purpose alone.
+"""
+module Timings
+
+using Core.Compiler: -, +, :, Vector, length, first, empty!, push!, pop!, @inline,
+    @inbounds, copy
+
+# What we record for any given frame we infer during type inference.
+function _typeinf_identifier(frame::Core.Compiler.InferenceState)
+    mi_info = (
+        frame.linfo,
+        frame.world,
+        copy(frame.sptypes),
+        frame.slottypes[2:end],  # Skip repeating the method instance
+    )
+    return mi_info
+end
+
+"""
+    Core.Compiler.Timing(mi_info, start_time, ...)
+
+Internal type containing the timing result for running type inference on a single
+MethodInstance.
+"""
+struct Timing
+    mi_info::Tuple{Core.MethodInstance,UInt64,Vector{Any},Vector{Any}}
```

Comment:
Now that this is no longer `Any`, I think it would be better to create a `struct` or `NamedTuple` for this type. The reason: meaning-by-position (i.e., `x[2]` is the method, `x[3]` is the type-tuple, ...) is neither transparent nor future-proof. Aside from avoiding stupid bugs, if we use a `struct` or `NamedTuple` SnoopCompile can also support multiple Julia (future) versions which make different choices, just by using `hasfield(Core.Compiler.Timing.InferenceInstanceInfo, :foo)`. In general I find multiversion support using `hasmethod` and `hasfield` to be more robust than `if Base.VERSION >= v"X.Y.Z"`, particularly while *developing* new features,  since version-checks only work for merged PRs.

I'm not attached to `InferenceInstanceInfo` as a name, use what you think is best.

Reply:
Thanks good suggestion! I've gone with `InferenceFrameInfo`, which seems to capture everything we mean by it?

Lemme know what you think, and if you have any other suggestions :) Thanks!